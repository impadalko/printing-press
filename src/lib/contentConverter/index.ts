import * as HTML from './HTML'

class ParsedLine {
  content = ''
  quoteDepth = 0

  constructor(line: string, quoteDepth = Infinity) {
    if (line === undefined) return
    this.content = line
    this.quoteDepth = 0
    while (this.content.startsWith('>') && this.quoteDepth < quoteDepth) {
      this.content = this.content.replace(/>\s?/, '')
      this.quoteDepth++
    }
  }
}

const getNextHTMLItem = (
  lines: string[],
  index: number,
  quoteDepth = 0
): [HTML.OptionalHTMLItem, number] => {
  if (index >= lines.length) return [null, index]
  let line = new ParsedLine(lines[index])
  if (line.quoteDepth < quoteDepth) return [null, index]
  if (line.quoteDepth > quoteDepth) {
    const quote = new HTML.Quote()
    while (line.quoteDepth > quoteDepth) {
      let subItem: HTML.OptionalHTMLItem
      ;[subItem, index] = getNextHTMLItem(lines, index, quoteDepth + 1)
      quote.extend(subItem)
      line = new ParsedLine(lines[index])
    }
    return [quote, index]
  }
  if (line.content === '') return [null, index + 1]
  if (line.content.startsWith('# ')) return [new HTML.H1(line.content), index + 1]
  if (line.content.startsWith('## ')) return [new HTML.H2(line.content), index + 1]
  if (line.content.startsWith('### ')) return [new HTML.H3(line.content), index + 1]
  if (line.content.startsWith('#### ')) return [new HTML.H4(line.content), index + 1]
  if (line.content.startsWith('##### ')) return [new HTML.H5(line.content), index + 1]
  if (line.content.startsWith('###### ')) return [new HTML.H6(line.content), index + 1]
  if (line.content === '---') return [new HTML.HR(), index + 1]
  if (line.content === '```') {
    const code = new HTML.Code()
    line = new ParsedLine(lines[++index], quoteDepth)
    while (index < lines.length && line.content != '```') {
      code.extend(line.content)
      line = new ParsedLine(lines[++index], quoteDepth)
    }
    return [code, index + 1]
  }
  const p = new HTML.P(line.content)
  index++
  while (index < lines.length) {
    // This can become inneficient on badly formatted texts
    const [subItem, nextIndex] = getNextHTMLItem(lines, index, quoteDepth)
    if (subItem instanceof HTML.P) {
      p.extend(subItem)
      index = nextIndex
    } else {
      return [p, index]
    }
  }
  return [p, index]
}

export const convert = (lines: string[]): string => {
  let converted = ''
  let nextHTMLItem: HTML.OptionalHTMLItem = null
  let index = 0
  while (index < lines.length) {
    // eslint-disable-next-line @typescript-eslint/no-extra-semi
    ;[nextHTMLItem, index] = getNextHTMLItem(lines, index)
    if (nextHTMLItem !== null) converted += nextHTMLItem.getParsed() + '\n'
  }
  return converted
}
