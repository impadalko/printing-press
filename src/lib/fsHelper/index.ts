import { promises as fs } from 'fs'
import path from 'path'

export const walk = async (inputPath: string): Promise<string[]> => {
  const dirContent = await fs.readdir(inputPath, { withFileTypes: true })
  let files: string[] = []
  for (const entry of dirContent) {
    const entryPath = path.join(inputPath, entry.name)
    if (entry.isDirectory()) files = files.concat(await walk(entryPath))
    else files.push(entryPath)
  }
  return files
}

export const ensureDir = async (filePath: string): Promise<void> => {
  await fs.mkdir(path.dirname(filePath), { recursive: true })
}

export const exists = async (filePath: string): Promise<boolean> => {
  try {
    await fs.access(filePath)
    return true
  } catch {
    return false
  }
}
