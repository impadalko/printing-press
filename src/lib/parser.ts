import fs from 'fs'
import readline from 'readline'
import YAML from 'yaml'

export interface ParsedFile {
  header: unknown;
  content: string[];
}

class ParsingError extends Error {
  constructor(msg: string) {
    super(`Could not parse: ${msg}`)
    Error.captureStackTrace(this, ParsingError)
  }
}

export const parseFile = async (path: string, headerDelimiter = '+++'): Promise<ParsedFile> => {
  const fileStream = fs.createReadStream(path)
  const linesInterface = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity
  })

  const lines = []
  for await (const line of linesInterface)
    lines.push(line)

  let header = ''
  const closedHeader = lines.some((line, index) => {
    if(index === 0 && line !== headerDelimiter)
      throw new ParsingError('First line must be a header delimiter')
    if(index !== 0 && line === headerDelimiter)
      return true
    if(line !== headerDelimiter){
      header += `${line}\n`
      return false
    }
  })

  if(!closedHeader)
    throw new ParsingError('Header was never closed')

  let parsedHeader = null
  try {
    parsedHeader = YAML.parse(header)
  } catch (e) {
    throw new ParsingError('Header is not a valid YAML')
  }

  let delimiterCount = 0
  const content: string[] = []
  lines.forEach((line) => {
    if (delimiterCount < 2){
      if (line === headerDelimiter)
        delimiterCount++
    }
    else
      content.push(line)
  })

  return {
    header: parsedHeader,
    content: content
  }
}
