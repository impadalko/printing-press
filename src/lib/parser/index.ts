import fs from 'fs'
import readline from 'readline'

type Header = { [k: string]: string }

interface ParsedFile {
  header: Header | null
  content: string[]
}

class ParsingError extends Error {
  constructor(msg: string) {
    super(`Could not parse: ${msg}`)
    Error.captureStackTrace(this, ParsingError)
  }
}

const parseHeader = (header: string): Header | null => {
  if (!header) return null
  const lines = header.split('\n')
  const parsedHeader: Header = {}
  lines.forEach((line) => {
    if (!line) return
    const splittedLine = line.split(':')
    if (splittedLine.length !== 2) throw new ParsingError(`Line "${line}" is not valid in header`)
    parsedHeader[splittedLine[0].trim()] = splittedLine[1].trim()
  })
  return parsedHeader
}

export const parseFile = async (path: string, headerDelimiter = '+++'): Promise<ParsedFile> => {
  const fileStream = fs.createReadStream(path)
  const lines = readline.createInterface({
    input: fileStream,
    crlfDelay: Infinity,
  })

  let delimiterCount = 0
  let header = ''
  const content: string[] = []
  for await (const line of lines) {
    if (delimiterCount === 0 && line !== headerDelimiter)
      throw new ParsingError('First line must be a header delimiter')
    if (delimiterCount < 2) {
      if (line === headerDelimiter) delimiterCount++
      else header += `${line}\n`
      continue
    }
    content.push(line)
  }

  if (delimiterCount === 1) throw new ParsingError('Header was never closed')

  const parsedHeader = parseHeader(header)
  return {
    header: parsedHeader,
    content: content,
  }
}
