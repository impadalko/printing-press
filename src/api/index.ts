import { promises as fs } from 'fs'
import path from 'path'

import { convert } from '../lib/contentConverter'
import { walk, ensureDir, exists } from '../lib/fsHelper'
import { parseFile } from '../lib/parser'

export interface BuildOptions {
  /** Path to the (optional) default template (relative to template path) */
  defaultTemplate?: string
}

/**
 * Build your website content
 * @param templatePath - Path to template folder
 * @param contentPath - Path to content folder
 * @param outputPath - Path to output folder
 * @param buildOptions - Options for the build process
 */
const buildContent = async (
  templatePath: string,
  contentPath: string,
  outputPath: string,
  buildOptions: BuildOptions = {}
): Promise<void> => {
  const contentFiles = await walk(contentPath)
  const loadedTemplates: { [key: string]: string } = {}
  const delimiterRegex = /<%= ([^%>]*) %>/g

  await Promise.all(
    contentFiles.map(async (file) => {
      const parsedFile = await parseFile(file)

      const template = parsedFile.header?.template || buildOptions.defaultTemplate
      if (!template) return

      if (!loadedTemplates[template]) {
        const templateFilePath = path.join(templatePath, template)
        if (!(await exists(templateFilePath))) {
          console.log(`${file}: Cannot find ${template} in template files`)
          return
        }
        loadedTemplates[template] = await fs.readFile(templateFilePath, { encoding: 'utf-8' })
      }

      let output = loadedTemplates[template]
      let match: RegExpExecArray | null = null
      while ((match = delimiterRegex.exec(loadedTemplates[template]))) {
        if (match[1] === 'content') output = output.replace(match[0], convert(parsedFile.content))
        else if (parsedFile.header?.[match[1]])
          output = output.replace(match[0], parsedFile.header?.[match[1]])
        else console.log(`${file}: Could not find ${match[1]} in header`)
      }

      const outputFilePath = path
        .join(outputPath, path.relative(contentPath, file))
        .replace(/\.[^.]+$/, '.html')
      await ensureDir(outputFilePath)
      await fs.writeFile(outputFilePath, output)
    })
  )
}

/**
 * Build your website
 * @param templatePath - Path to template folder
 * @param contentPath - Path to content folder
 * @param outputPath - Path to output folder
 * @param publicPath - Path to public folder
 * @param buildOptions - Options for the build process
 */
const build = async (
  templatePath: string,
  contentPath: string,
  outputPath: string,
  publicPath?: string,
  buildOptions: BuildOptions = {}
): Promise<void> => {
  await buildContent(templatePath, contentPath, outputPath, buildOptions)

  if (publicPath) {
    const publicFiles = await walk(publicPath)
    await Promise.all(
      publicFiles.map(async (src) => {
        const dest = path.join(outputPath, path.relative(publicPath, src))
        await ensureDir(dest)
        await fs.copyFile(src, dest)
      })
    )
  }
}

export default {
  build,
  buildContent,
}
