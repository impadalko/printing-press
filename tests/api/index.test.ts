import { promises as fs } from 'fs'
import path from 'path'

import api from '../../src/api'
import { walk } from '../../src/lib/fsHelper'

const removeExt = (file: string): string => {
  return file.replace(/\.[^.]+$/, '')
}

const templatePath = `${__dirname}/templates`
const contentPath = `${__dirname}/content`
const outputPath = `${__dirname}/output`
const publicPath = `${__dirname}/public`
const defaultTemplate = 'default.html'

afterEach(() => {
  return fs.rm(outputPath, { recursive: true, force: true })
})

test('Built files have the same file structure as the content files', async () => {
  await api.buildContent(templatePath, contentPath, outputPath, { defaultTemplate })

  const inputFiles = new Set(
    (await walk(contentPath)).map(removeExt).map((e) => path.relative(contentPath, e))
  )
  const outputFiles = new Set(
    (await walk(outputPath)).map(removeExt).map((e) => path.relative(outputPath, e))
  )
  expect(inputFiles).toEqual(outputFiles)
})

test('Built files have the correct content', async () => {
  await api.buildContent(templatePath, contentPath, outputPath, { defaultTemplate })

  const outputFiles = await walk(outputPath)
  for (const output of outputFiles) {
    const content = await fs.readFile(output, { encoding: 'utf-8' })
    expect(content).toMatchSnapshot()
  }
})

test('Public folder files are copied to the output folder', async () => {
  await api.build(templatePath, contentPath, outputPath, publicPath)

  const outputFiles = new Set((await walk(outputPath)).map((e) => path.relative(outputPath, e)))
  const publicFiles = (await walk(publicPath)).map((e) => path.relative(publicPath, e))
  for (const publicFile of publicFiles) {
    expect(outputFiles.has(publicFile)).toBe(true)
  }
})
