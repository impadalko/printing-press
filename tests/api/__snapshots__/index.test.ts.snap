// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Built files have the correct content 1`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <title>Test</title>
    <meta charset=\\"UTF-8\\" />
    <meta name=\\"viewport\\" content=\\"width=device-width,initial-scale=1\\" />
    <meta name=\\"description\\" content=\\"\\" />
  </head>
  <body>
    <p>Amazing example!</p>

  </body>
</html>
"
`;

exports[`Built files have the correct content 2`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <title>Another example</title>
    <meta charset=\\"UTF-8\\" />
    <meta name=\\"viewport\\" content=\\"width=device-width,initial-scale=1\\" />
    <meta name=\\"description\\" content=\\"\\" />
  </head>
  <body>
    <h3>Amazing example!</h3>

  </body>
</html>
"
`;

exports[`Built files have the correct content 3`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <title>Title</title>
    <meta charset=\\"UTF-8\\" />
    <meta name=\\"viewport\\" content=\\"width=device-width,initial-scale=1\\" />
    <meta name=\\"description\\" content=\\"\\" />
  </head>
  <body>
    <h1>Unreachable</h1>
    <p>This should never run:</p>
<pre><code class=\\"python\\">
while(False):
  print('how did you get here?')
</code></pre>

    <p>Please desconsider the previous examples</p>
  </body>
</html>
"
`;

exports[`Built files have the correct content 4`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <title>Complex markdown</title>
    <meta charset=\\"UTF-8\\" />
    <meta name=\\"viewport\\" content=\\"width=device-width,initial-scale=1\\" />
    <meta name=\\"description\\" content=\\"\\" />
  </head>
  <body>
    <h2>Sub-title</h2>
<blockquote><p>Quote</p>
<blockquote><p>Nested quote</p></blockquote>
<p>Top level quote</p></blockquote>
<ul>
<li>list</li>
<li>
still the same list
<ul>
<li>nested list</li>
</ul>
<ol>
<li>nested ordered list</li>
<li>another element</li>
</ol>
</li>
<li>back to first list</li>
</ul>
<h2>Another sub-title</h2>
<p>Test test</p>

  </body>
</html>
"
`;

exports[`Built files have the correct content 5`] = `
"<!DOCTYPE html>
<html lang=\\"en\\">
  <head>
    <title>Example</title>
    <meta charset=\\"UTF-8\\" />
    <meta name=\\"viewport\\" content=\\"width=device-width,initial-scale=1\\" />
    <meta name=\\"description\\" content=\\"\\" />
  </head>
  <body>
    <p>Another <em>example</em></p>
<blockquote><p>This is just an <a href=\\"http://example.com\\">example</a></p>
<p><code>Lorem Ipsum</code></p></blockquote>

  </body>
</html>
"
`;

exports[`Files with draft in the header are ignored 1`] = `
Array [
  "b.html",
]
`;
